### PR TITLE
feat(proxy): dynamically parse retry response from body

### PR DESCRIPTION
**Pull request by hassan254-prog**
## Describe the problem and your solution 

- Dynamically parse retry response from body

---

**Add Dynamic Retry Parsing from Response Body to Proxy Layer**

This PR adds support for extracting retry timing information (e.g., when to retry after rate limit) from the HTTP response body, in addition to headers, within the proxy service. The implementation generalizes retry logic to handle both header- and body-based strategies, adds schema and type support for configuring body-based retry detection, updates the Google provider config as an example, and extends unit tests to thoroughly cover new parsing scenarios (ISO, epoch, and relative delays).

**Key Changes:**
• Refactored proxy retry decision flow to support dynamic retry extraction from response bodies using configurable regex and ``JSON`` path.
• Introduced a new `proxy.retry.`in_body`` configuration option to ``JSON`` schema, types, and providers.
• Updated Google provider to specify retry parsing from body as an applied example.
• Enhanced test coverage with new unit tests for various body parsing strategies and edge cases.
• Added lodash-es/get to handle safe access for ``JSON`` path in body parsing.

**Affected Areas:**
• Shared proxy retry logic (packages/shared/lib/services/proxy/retry.ts)
• Provider configuration files and validation schema
• Types and ``API`` interface for proxy configuration
• Google provider's proxy section
• Unit test suite for retry logic

**Potential Impact:**

**Functionality**: Enables broader auto-retry support for APIs that encode retry timing in body, increasing reliability for providers (such as Google). Maintains backward compatibility by extending, not altering, existing header logic.

**Performance**: Minimal; regex and JSON path evaluations are straightforward, but additional parsing logic is executed on relevant error responses.

**Security**: No direct issues; regex parsing may be error-prone but is confined to error handling.

**Scalability**: No negative impact; logic is per-request and low-overhead.

**Review Focus:**
• Ensure correctness and robustness of `getRetryFromBody` and `parseRetryValue` logic for various edge cases (epoch, ``ISO``, malformed, etc).
• Review regular expression usage and configuration path mapping for body parsing.
• Verify schema and types align with actual provider configurations.
• Check for full test coverage and meaningful error reporting in ambiguous cases.

<details>
<summary><strong>Testing Needed</strong></summary>

• Unit tests are comprehensive, but real-world ``API`` integration tests with providers sending retry info in body are suggested.
• Manually verify the Google provider handling of rate limit body messages.
• Check that legacy header-based retry logic continues to function as expected.

</details>

<details>
<summary><strong>Code Quality Assessment</strong></summary>

**packages/shared/lib/services/proxy/retry.ts**: Refactored for clarity and extensibility; modular helper methods improve maintainability. Some review comments request further clarity in error reason strings and prefixed error reasons, which were addressed.

**packages/shared/lib/services/proxy/retry.unit.test.ts**: Thorough test coverage with good organization and readable scenario names.

**scripts/validation/providers/schema.json**: Schema extended in a backward-compatible and well-documented way.

**packages/types/lib/proxy/api.ts**: Types extended cleanly.

**packages/providers/providers.yaml**: Provider config extended with clear strategy demonstration (Google).

</details>

<details>
<summary><strong>Best Practices</strong></summary>

**Testing**:
• Strong unit testing for new code paths.
• Regression testing for header-based logic preserved.

**Extensibility**:
• Retry logic now modular for both headers and body (future-proof).

**Type Safety**:
• Types and schemas are updated alongside logic.

</details>

<details>
<summary><strong>Possible Issues</strong></summary>

• Regex configuration for body parsing must match provider responses exactly—misconfiguration could silently break retry inference.
• Non-string JSON fields at the given path, or unexpected error messages, may cause failed retry detections.
• Performance could be minimally affected if regexes or JSON parsing become complex.

</details>

---
*This summary was automatically generated by @propel-code-bot*

### CODE REVIEW

File: packages/shared/lib/services/proxy/retry.ts [hunk 1]
```diff
@@ -94,35 +107,106 @@ export function getRetryFromHeader({
...
+}): ReturnType<typeof parseRetryValue> {
+    const responseBody = err.response?.data;
+    if (!responseBody) {
+        return { found: false, reason: 'in_body:no_response' };
+    }
+
+    const rawMessage = get(responseBody, retryPath);
+    if (!rawMessage || typeof rawMessage !== 'string') {
+        return { found: false, reason: 'in_body:path_not_found' };
+    }
```
Comment (bodinsamuel) [inline packages/shared/lib/services/proxy/retry.ts]: Can you split the two reason just for clarity

File: packages/providers/providers.yaml [hunk 2]
```diff
@@ -4501,6 +4501,11 @@ google:
         access_type: offline
         prompt: consent
     proxy:
+        retry:
+            in_body:
+                path: error.message
+                value: User-rate limit exceeded.  Retry after ${value}
```
Comment (bodinsamuel) [inline packages/providers/providers.yaml]: I feel like this prop could just be a regex instead of this custom syntax
```suggestion
                value: /User-rate limit exceeded.  Retry after (.+)/
```

File: packages/shared/lib/services/proxy/retry.ts [hunk 3]
```diff
@@ -94,35 +107,106 @@ export function getRetryFromHeader({
...
-        const retryHeaderVal = err.response?.headers[retryHeader] || err.response?.headers[retryHeader.toLowerCase()];
-
-        if (!retryHeaderVal) {
-            return { found: false, reason: 'after:no_header' };
+    if (type === 'after') {
+        if (!rawValue) {
+            return { found: false, reason: `${type}:no_${source}` };
         }
 
-        // TODO: handle non-seconds header (e.g: linear)
```
Comment (bodinsamuel) [inline packages/shared/lib/services/proxy/retry.ts:124]: can you put the comment again please

File: packages/shared/lib/services/proxy/retry.ts [hunk 4]
```diff
@@ -94,35 +107,106 @@ export function getRetryFromHeader({
...
+        } else {
+            const dateValue = new Date(rawValue).getTime();
+            if (isNaN(dateValue)) {
+                return { found: false, reason: `${prefix}_invalid_date` };
+            }
+            retryAtEpoch = Math.floor(dateValue / 1000);
+        }
 
         if (retryAtEpoch <= currentEpochTime) {
             return { found: false, reason: 'at:invalid_wait' };
```
Comment (bodinsamuel) [inline packages/shared/lib/services/proxy/retry.ts]: missing prefix

File: packages/shared/lib/services/proxy/retry.ts [hunk 5]
```diff
@@ -94,35 +107,106 @@ export function getRetryFromHeader({
...
+    source: 'header' | 'body';
+}): ReturnType<typeof getRetryFromHeader> {
+    const prefix = `${type}:${source}`;
+
     if (type === 'at') {
-        const resetTimeEpoch = err.response?.headers[retryHeader] || err.response?.headers[retryHeader.toLowerCase()];
-        if (!resetTimeEpoch) {
-            return { found: false, reason: 'at:no_header' };
+        if (!rawValue) {
+            return { found: false, reason: `${type}:no_${source}` };
```
Comment (bodinsamuel) [inline packages/shared/lib/services/proxy/retry.ts]: this if could be put once at the beginning

File: packages/shared/lib/services/proxy/retry.ts [hunk 6]
```diff
@@ -94,35 +107,106 @@ export function getRetryFromHeader({
...
+                return { found: false, reason: `${prefix}_invalid_date` };
+            }
+            retryAtEpoch = Math.floor(dateValue / 1000);
+        }
 
         if (retryAtEpoch <= currentEpochTime) {
             return { found: false, reason: 'at:invalid_wait' };
         }
 
-        // TODO: handle non-seconds header
```
Comment (bodinsamuel) [inline packages/shared/lib/services/proxy/retry.ts:113]: can you put this back please

**Review by bodinsamuel — COMMENTED** [review]
Nice, definitely useful 🚀 
Again, don't forget to put some context on your PR. Linear might come and go, not the git commit

File: packages/shared/lib/services/proxy/retry.ts (hunk 5 above)
Comment (hassan254-prog) [reply packages/shared/lib/services/proxy/retry.ts]: yes, thanks updated in the latest commit 🙏 .

File: packages/shared/lib/services/proxy/retry.ts (hunk 4 above)
Comment (hassan254-prog) [reply packages/shared/lib/services/proxy/retry.ts]: thanks 🙏

File: packages/providers/providers.yaml (hunk 2 above)
Comment (hassan254-prog) [reply packages/providers/providers.yaml]: updated to the new regex pattern matching logic in the latest commit  🚀

**Review by bodinsamuel — APPROVED** [review]
